### PR TITLE
Tm

### PR DESCRIPTION
**Pull request by toddr**

### CODE REVIEW

File: t/SATest.pm [hunk 1]
```diff
@@ -372,10 +379,13 @@ sub salearnrun {
...
   (-d "log/d.$testname") or mkdir ("log/d.$testname", 0755);
-  system ("$salearnargs > log/d.$testname/${Test::ntest}");
+
+  my $test_number = test_number();
+
+  system ("$salearnargs > log/d.$testname/$test_number");
   $salearn_exitcode = ($?>>8);
   if ($salearn_exitcode != 0) { return undef; }
-  &checkfile ("d.$testname/${Test::ntest}", $read_sub) if (defined $read_sub);
+  &checkfile ("d.$testname/$test_number", $read_sub) if (defined $read_sub);
```
Comment (atoomic) [inline t/SATest.pm:388]: probably better preserving the braces around ${test_number} ( here an all other locations )

File: t/autolearn.t [hunk 2]
```diff
@@ -2,16 +2,15 @@
 
 use lib '.'; use lib 't';
 use SATest; sa_t_init("autolearn");
-use Test;
 
-use constant TEST_ENABLED => eval { require DB_File; };
+use constant HAS_DB_FILE => eval { require DB_File; };
 
+use Test::More;
 BEGIN {
```
Comment (atoomic) [inline t/autolearn.t]: is that BEGIN block required ?

File: t/autolearn_force.t [hunk 3]
```diff
@@ -2,16 +2,15 @@
 
 use lib '.'; use lib 't';
 use SATest; sa_t_init("autolearn_force");
-use Test;
 
-use constant TEST_ENABLED => eval { require DB_File; };
+use constant HAS_DB_FILE => eval { require DB_File; };
 
+use Test::More;
 BEGIN {
```
Comment (atoomic) [inline t/autolearn_force.t]: same question here, is it necessary to skip test in the BEGIN block ?

File: t/autolearn_force_fail.t [hunk 4]
```diff
@@ -2,17 +2,15 @@
...
 use SATest; sa_t_init("autolearn_force_fail");
-use Test; 
 
-use Test;
-
-use constant TEST_ENABLED => eval { require DB_File; };
+use constant HAS_DB_FILE => eval { require DB_File; };
 
+use Test::More;
 BEGIN {
```
Comment (atoomic) [inline t/autolearn_force_fail.t]: same question for BEGIN block

File: t/bayesdbm_flock.t [hunk 5]
```diff
@@ -17,10 +16,12 @@ BEGIN {
     unshift(@INC, '../blib/lib');
   }
 
-  plan tests => ((TEST_ENABLED && !$RUNNING_ON_WINDOWS) ? 48 : 0);
-};
+  plan skip_all => "Long running tests disabled" unless conf_bool('run_long_tests');
+  plan skip_all => "Tests don't work on windows" if $RUNNING_ON_WINDOWS;
```
Comment (atoomic) [inline t/bayesdbm_flock.t]: not sure at first glance where comes from the variable $RUNNING_ON_WINDOWS 
is it defined? maybe add a comment there?

File: t/bayesdbm_flock.t [hunk 6]
```diff
@@ -3,11 +3,10 @@
...
-use Test;
 
-use constant TEST_ENABLED => conf_bool('run_long_tests') &&
-                        eval { require DB_File; };
+use constant HAS_DB_FILE => eval { require DB_File };
 
+use Test::More;
 BEGIN { 
   if (-e 't/test_dir') {
     chdir 't';
```
Comment (atoomic) [inline t/bayesdbm_flock.t:12]: would convert this check to be oneliner with postfix check

File: t/bayessdbm_seen_delete.t [hunk 7]
```diff
@@ -3,11 +3,10 @@
 use Data::Dumper;
```
Comment (atoomic) [inline t/bayessdbm_seen_delete.t:3]: note sure where Data::Dumper is used but using Test::More helper explain instead might be a good idea

File: t/config_errs.t [hunk 8]
```diff
@@ -1,13 +1,10 @@
...
-# 'make test'.
+use Test::More;
 
-use Test;
-
-plan tests => 0;
-exit 0;     
+# TODO: Now we're using Test::More, we could choose to make the failing tests TODO tests so we know if they start passing.
+plan skip_all => q{This script does not work yet. There are (a) lots of failures and (b) the 'plan' line means all tests are effectively ignored from 'make test'.};
+exit 0;
```
Comment (atoomic) [inline t/config_errs.t:7]: seems like exit is not required after a plan skip_all ?

File: t/config_text.t [hunk 9]
```diff
@@ -1,14 +1,17 @@
...
 use lib '.'; use lib 't';
 use SATest; sa_t_init("config_text");
 
 # skip the test on Windows; the switches contain spaces, which does not
 # work too well with win32 CMD.EXE
-use Test; BEGIN { plan tests => IS_WINDOWS ? 0 : 2 };
-exit if IS_WINDOWS;
+
+use Test::More;
+BEGIN {
```
Comment (atoomic) [inline t/config_text.t]: is BEGIN really required there?

File: t/db_awl_path.t [hunk 10]
```diff
@@ -36,8 +36,8 @@ my $error = do {
   <$fh>;
 };
 
-print "# $error\n";
-ok($error, qr/(cannot create tmp lockfile)|(unlink of lock file.*failed)/, "Check we get the right error back");
+diag $error;
+like($error, qr/(cannot create tmp lockfile)|(unlink of lock file.*failed)/, "Check we get the right error back");
```
Comment (atoomic) [inline t/db_awl_path.t:40]: minor improvement to consider use like .. or diag ...

File: t/mimeparse.t [hunk 11]
```diff
@@ -107,13 +108,6 @@ my $sa = Mail::SpamAssassin->new({
     dont_copy_prefs   => 1,
 });
 
-my $numtests = 5;
```
Comment (atoomic) [inline t/mimeparse.t:110]: maybe could have preserve this dynamic rule to set the total number of tests

File: t/mkrules.t [hunk 12]
```diff
@@ -455,17 +455,21 @@ sub mkrun {
 
   my $scrargs = "$perl_path -I../lib ../build/mkrules $args";
   print ("\t$scrargs\n");
-  system ("$scrargs > log/$testname.${Test::ntest} $post_redir");
+
+  my $test_number = test_number();
```
Comment (atoomic) [inline t/mkrules.t:459]: this is tricky as test_number is auto imported by "use SATest;" but fine

File: t/recreate.t [hunk 13]
```diff
@@ -3,7 +3,7 @@
 use Data::Dumper;
 use lib '.'; use lib 't';
 use SATest; sa_t_init("recreate");
-use Test;
+use Test::More tests => 9;
 
 BEGIN {
```
Comment (atoomic) [inline t/recreate.t:8]: this BEGIN block seems to be a boilerplate, would consider an helper setup_inc from SATest (similar to the one used by the perl testsuite)

File: t/root_spamd.t [hunk 14]
```diff
@@ -4,14 +4,14 @@
...
-use Test;
 
-use constant TEST_ENABLED => conf_bool('run_root_tests');
-use constant IS_ROOT => eval { ($> == 0); };
-use constant RUN_TESTS => (TEST_ENABLED && IS_ROOT);
-
-BEGIN { plan tests => (RUN_TESTS ? 14 : 0) };
-exit unless RUN_TESTS;
+use Test::More;
+BEGIN {
```
Comment (atoomic) [inline t/root_spamd.t]: not sure the BEGIN block is required there
regarding the number of tests using this pattern would also consider a helper like skip_if_cannot_run_tests_as_root() to always perform these two checks

File: t/spamc.t [hunk 15]
```diff
@@ -2,9 +2,14 @@
 
 use lib '.'; use lib 't';
 use SATest; sa_t_init("spamc");
-use Test; plan tests => ($NO_SPAMC_EXE ? 0 : 2);
 
-exit if $NO_SPAMC_EXE;
+use Test::More;
+BEGIN {
+  plan skip_all => "No SPAMC exe" if $SKIP_SPAMC_TESTS;
```
Comment (atoomic) [inline t/spamc.t]: cannot find where SKIP_SPAMC_TESTS is set ? is it a config variable?

File: t/spamd_client.t [hunk 16]
```diff
@@ -19,26 +19,28 @@ if (-e 'test_dir') {            # running from test directory, not ..
...
-  $num_tests += 21;
-}
-
-use Test; plan tests => ($DO_RUN ? $num_tests : 0);
+use Test::More;
+BEGIN {
+  plan skip_all => "Spamd tests disabled" if $SKIP_SPAMD_TESTS;
+  plan skip_all => "Long running tests disabled" unless conf_bool('run_long_tests');
+  
+  # TODO: These should be skips down in the code, not changing the test count.
```
Comment (atoomic) [inline t/spamd_client.t]: yeap ++

File: t/spamd_prefork_stress_3.t [hunk 17]
```diff
@@ -2,19 +2,17 @@
...
 use SATest; sa_t_init("spamd_prefork_stress_3");
-use Test;
 
-our $RUN_THIS_TEST;
-BEGIN {
-  $RUN_THIS_TEST = conf_bool('run_spamd_prefork_stress_test') &&
-                        conf_bool('run_long_tests');
-  plan tests => ($SKIP_SPAMD_TESTS || !$RUN_THIS_TEST ? 0 : 291)
+use Test::More;
+ BEGIN {
```
Comment (atoomic) [inline t/spamd_prefork_stress_3.t]: is the BEGIN block required ? regarding the exit which were used to be there at run time pretty sure it s not required

File: t/uribl_all_types.t [hunk 18]
```diff
@@ -5,15 +5,14 @@
...
 
-use constant TEST_ENABLED => conf_bool('run_net_tests') && conf_bool('run_long_tests');
-use constant DO_RUN => TEST_ENABLED && can_use_net_dns_safely();
-use Test;
-
+use Test::More;
 BEGIN {
-  plan tests => (DO_RUN ? 3 : 0),
-};
+  plan skip_all => "Long running tests disabled" unless conf_bool('run_long_tests');
```
Comment (atoomic) [inline t/uribl_all_types.t]: bonus: helper like
skip_all_if_long_tests_disabled()
skip_all_if_net_tests_disabled()
skip_all_if_cannot_use_dns()

**Review by atoomic — COMMENTED** [review]
looks good, wonder if some of the BEGIN block are really necessary when setting the plan or skipping all tests
Also introducing some extra helpers in SATests might be a good idea and would avoid boilerplate

File: t/autolearn.t (hunk 2 above)
Comment (toddr) [reply t/autolearn.t]: No. it's more of a pattern. There are cases where it's needed to match previous behavior.

File: t/db_awl_path.t (hunk 10 above)
Comment (toddr) [reply t/db_awl_path.t:40]: Fixed in later commit.

**Pull request closed without merging**